### PR TITLE
[RFC] CI: Make ruff flag imports from forbidden modules.

### PR DESCRIPTION
**Pull request by PapyChacal**
RFC Because I'm not sure if we're interested in doing it this way?

We try in reviews to prevent imports from some modules (`xdsl.parser.X`, `xdsl.irdl.irdl` and `xdsl.ir.core`, to my knowledge. please tell me if I missed some.)
It feels to me like something we want a tool to check.

Ruff has a rule for it.

- CON: No auto-fix provided. it is the developer's responsibility to fix it.
- PRO: Well, it automatically finds them. Also allows for a clear to-the-point error message to not make the user scratch his/her head too much.

Also note that for those cases, it's always only about removing the last nested module from the from, it's really not cumbersome, for what it's worth. It then often leads to misformatted imports, but *those* are then auto-fixed.

WDYT?

### CODE REVIEW

**Comment by webmiche** [comment]
That sounds like a great opportunity for such a tool, I like it!

**Comment by webmiche** [comment]
What does the error message look like?

**Comment by PapyChacal** [comment]
> What does the error message look like?
```
xdsl/parser/core.py:11:1: TID251 `xdsl.ir.core` is banned: Use xdsl.ir instead.
```
so the line is clickable on VSCode - and I guess most IDEs ? - to go directly to the line and erase the problematic bit ore copy paste the mentionned module. :slightly_smiling_face:

**Comment by PapyChacal** [comment]
> > What does the error message look like?
> 
> ```
> xdsl/parser/core.py:11:1: TID251 `xdsl.ir.core` is banned: Use xdsl.ir instead.
> ```
> 
> so the line is clickable on VSCode - and I guess most IDEs ? - to go directly to the line and erase the problematic bit ore copy paste the mentionned module. 🙂

Nit: It could look better with backquotes on the advised module too?

**Review by superlopuh — APPROVED** [review]
👍👍👍

File: pyproject.toml [hunk 1]
```diff
@@ -37,6 +37,14 @@ ignore = [
 line-length = 300
 target-version = "py310"
 
+[tool.ruff.flake8-tidy-imports.banned-api]
+"xdsl.parser.core".msg = "Use xdsl.parser instead."
+"xdsl.parser.attribute_parser".msg = "Use xdsl.parser instead."
+"xdsl.parser.affine_parser".msg = "Use xdsl.parser instead."
+"xdsl.ir.core".msg = "Use xdsl.ir instead."
+"xdsl.irdl.irdl".msg = "Use xdsl.irdl instead"
```
Comment (compor) [inline pyproject.toml:45]: This is great.

**Review by compor — APPROVED** [review]

**Pull request merged**